ClangManager, project template fix and others

gudaol opened the pull request:
Added CLangManager, fixed project model
Added ClangSharp + CLangManager basic functionality, multithreading from ThreadPool
Added support for clang, clang++ compilers, selectable from GUI
Fixed project, project configuration initializations
Fixed missing GetTypeTags
Refactored CBinding.CompileTarger -> now uses built in CompileTarget class
Segfault on project creation - load, unknown cause yet

Commit:
Added CLangManager, fixed project model
Added ClangSharp + CLangManager basic functionality, multithreading from ThreadPool
Added support for clang, clang++ compilers, selectable from GUI
Added support for selectable C/C++ standard, selectable from GUI
Fixed project, project configuration initializations
Fixed missing OnGetTypeTags, now File creation dialog works with C/C++
Refactored CBinding.CompileTarger -> now uses built in CompileTarget class
Segfault on project creation - load, unknown cause yet

Commit:
Fixed accidental revert

gudaol commented:
Tried to keep the commits as clean as possible, sorry for the GUI part, MD kept regenerating them.

Commit:
Update README.md

Commit:
Fixed segfault on start, refactoring
Cause was an infinite loop in OnGetReferencedItems, which called base.GetReferencedItems, which called.... "derived".OnGetRefrencedItems!
Refactored Target cproj ItemProperty to OutputType to correspond with csproj and to avoid later confusion with Build targets

Commit:
Merge branch 'master' of https://github.com/gudaol/cbinding

Commit:
Added ClangSharp.dll, libclang.so to the repo & project

Commit:
Merge branch 'master' of https://github.com/gudaol/cbinding

Commit:
Removed explicit dll, code completion basics
Removed DLL because the generated source file is... well full of errors
Added "Generated.cs, Extensions.cs" which are ClangSharp files - hand fixed CXCompletionString, CXCodeCompleteResult structures and related functions
Added very-very-very basic code completion feature, list is finally working but full of "garbage" (very unlikely suggestions)
Removed probably unneeded code from CTextEditorExtension
Fixed project file, which generated ClangSharp.dll... as documentation
Refactored CLangManager Add-UpdateTranslationUnit methods - a bit rationalized them

Commit:
Some rework of code completion
[TODO] Completions are now sorted by their clang-given priority - but sorting is ineffective somehow in the code completion dialog...
Removed debug/testing console writes
Added classes that will be probably used in the near future
>> completion unit - temporary, sortable collection of CompletionData-s
>> completion category -  to categorize suggestions
>> symbol database - might need to have an own AST container

Commit:
Symbol database, code completion
Added symbol database - when making translation units the database is filled with parsed information,  reworked Parser/*
Added Code completion improvements - completion category + parameter hinting lookup can provide the signature
Refactored ClangManager - from now not singleton, but instatiated for every project
Removed many unneeded files
CDocumentParser is WIP - removed from Compiled files and from CBinding.addin.xml

Commit:
Reworked completion
Fixed completion sorting
Fixed completion not working in C++ projects
Re-Fixed completion row-column offset
Refactored symbol database to instatiate for every file
Fixed bug that functions appeared more and more after every save

Commit:
Fixed C++ issue
Previous commit didn't contain cpp fix as intended, but this does

Commit:
Fixed completion invokement requirements

Commit:
Fixed unsaved file initialization
CC now works properly

Commit:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	CBinding/CBinding.csproj
	CBinding/packages.config

Commit:
Fix project to work with my config

Commit:
Symbol database extending
Initials to support * templates in future

Commit:
Code completion filtering
filters: operators, destructors and constructors

Commit:
Added "Go to declaration"
involved:
Added custom SourceLocation class
Fixed ClangSharp generated.cs

tweak:
UpdateTranslationUnits should be more efficient this way according to the clang documentation
http://clang.llvm.org/doxygen/group__CINDEX__TRANSLATION__UNIT.html#ga524e76bf2a809d037934d4be51ea448a

Commit:
Fixed left out symbol database building

Commit:
Find references (buggy) + initials
Find references works across files, but sometimes the results are misalignes by a fix offset (e.g.: 3), it seems clang gives these results falsely, couldnt find out yet why. Aside this it works!

Initials for Rename and Find derived classes/symbols

Commit:
Namespace refactoring

Commit:
Overhaul of parsing, Go to definition
Reworked: CLang now parses constantly the files if there is an unsaved/dirty Document opened in the IDE, in a background thread.
Reworked: Parsing is also more detailed, many fields added
Added: Go to definition command in context menu
Refactored: Go to declaration - more OO, has class now

Commit:
Parser improvements: on the fly errors
Improved parser, optimized in a few places - TODO: only parse on document contect change + nicen code
Fix needed: EFBBBF bíte order sequence misalignes clang cursor locations, need a workaround.

Updated references of roslyn branch which removed nrefactory6 and desktop - not used them anyways

Commit:
Removed debug messages + added missing lock

Commit:
Parsing improvements, fixes
Parsign is now invoked on-demand through CDocumentParser.Parse
Fixed Default configration change

Commit:
Added rename functionality
Rename: very basic, might be buggy
TODO: On the fly error markers: removing them is buggy, might crash GTK
TODO: Bug in parsing still persist: Unicode BOM chars misalign cursor placements

Commit:
Reworked parsing
TODO: Fix BOM misalignment in editor
TODO: Fix rename

Commit:
Fixes
Threading fixes
Rename fixed
Parameter hinting fixed

Commit:
Fixes
Fixed BOM misalignment problems
Improved rename to validate input, and warn if invalid

Commit:
Fixed Go to declaration
Go to declaration sometimes went to definition, fixed

Commit:
Added '(' to completion chars

Commit:
Documentation

Commit:
Merge pull request #1 from gudaol/unstable

Merge stable changes into master

mhutch commented on CBinding/CBinding.csproj:
```diff
@@ -106,57 +108,79 @@
-    <Reference Include="Microsoft.CodeAnalysis.CSharp.Desktop">
-      <HintPath>..\packages\Microsoft.CodeAnalysis.CSharp.1.0.0-rc2\lib\net45\Microsoft.CodeAnalysis.CSharp.Desktop.dll</HintPath>
+    <Reference Include="Mono.TextEditor">
+      <HintPath>..\..\monodevelop\main\build\bin\Mono.TextEditor.dll</HintPath>
```
You shouldn't need to reference `..\..\monodevelop\main\build\bin`, the AddinMaker targets do it automatically.

mhutch replied:
Bump :) Can't merge until this is addressed.

gudaol replied:
on my config it wont compile if i remove them, somehow addin maker targets wont work.

mhutch commented on CBinding/CBinding.csproj:
```diff
@@ -106,57 +108,79 @@
-    <Reference Include="Microsoft.CodeAnalysis.CSharp">
-      <HintPath>..\packages\Microsoft.CodeAnalysis.CSharp.1.0.0-rc2\lib\net45\Microsoft.CodeAnalysis.CSharp.dll</HintPath>
+    <Reference Include="MonoDevelop.AddinMaker">
+      <HintPath>..\..\MonoDevelop.AddinMaker\bin\Debug\MonoDevelop.AddinMaker.dll</HintPath>
```
You definitely shouldn't depend on the AddinMaker addin.

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,482 @@
+			project = proj;
+			index = clang.createIndex (0, 0);
+			translationUnits = new Dictionary<string, CXTranslationUnit> ();
+			project.DefaultConfigurationChanged += CompilerArgumentsUpdate;
```
Do you need to unsubscribe this when disposing so it doesn't leak? If this has the same lifetime as the project, it should be fine as-is, just checking.

gudaol replied:
CLangManager has the same lifetime as the project, but adding the unsubscribe just to be sure.

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,482 @@
+	/// </summary>
+	public class CLangManager : IDisposable
+	{
+		public readonly object syncroot = new object ();
```
This shouldn't be public.

gudaol replied:
In CParsedDocument i interact with translation units, had no better solution than to expose the lock. Ill figure something out.

mhutch replied:
Or if it is public, it should be called SyncRoot.

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,482 @@
+	public class CLangManager : IDisposable
+	{
+		public readonly object syncroot = new object ();
+		private CProject project;
```
Please leave off the redundant `private` accessibility modifier as per the MonoDevelop code guidelines: http://www.monodevelop.com/developers/#coding-guidelines

If you have source analysis enabled, it should hint about these (not 100% sure this rule was ported to Roslyn though).

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,482 @@
+
+		public Dictionary<string, CXTranslationUnit> TranslationUnits {
+			get {				
+				lock (syncroot) {
```
This lock is redundant - getting the field is atomic. You should probably add a comment about how the dictionary itself isn't threadsafe and accesses to the dictionary need to be locked on the SyncRoot. Or ideally hide this entirely behind threadsafe API, or use a ConcurrentDictionary? Or maybe switch to locking the dictionary, instead of a syncroot object?

gudaol replied:
I could not find any information about libclang's internal threading solutions, the best i could find were some stackoverflow topic and a short mailing list discussion, where only assumptions were made. This lock is needed because clangmanager handles getting cursor, cursor references and cursor between files sometimes are a mess and can only identified by its USR, and e.g. finding a cursor by its USR while reparsing is in progress in a file could be a fault.

Commit:
Implemented the changes required
Removed private modifiers
Proper Dispose
syncroot to SyncRoot and related documentation expanding

Commit:
Fixed huge performance drop
Refactored codecomplete not to depend on editor
Removed redundant this qualifiers

Commit:
Merge pull request #2 from gudaol/performance-and-nicen

Fixed huge performance drop

Commit:
Parsing operation is now cancellable
Little performance gain by cancelling parsings when it should be cancelled

Commit:
Merge pull request #4 from gudaol/cancellable

Parsing operation is now cancellable

Commit:
Ported Class pad + fixes

Fixed issues:
	parameter hinting multiple results
	rename and find ref fix

Commit:
Merge pull request #6 from gudaol/parserfixes-classpad

Ported Class pad + fixes

Commit:
Parameter hinting
Refixed find refs+rename
Param hinting is now nearly perfect - needs addition: determine function return type/variable type before hinted function to filter results with same name in different classes + visibility issues with same named functions (needs some thinking, clang is unable to do this it seems).

Commit:
Suuper-duper parameter hinting

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,528 @@
+		/// <returns>
+		/// A <see cref="CXTranslationUnit"/>: The Translation unit created
+		/// </returns>
+		public CXTranslationUnit createTranslationUnit(CProject proj, string fileName, CXUnsavedFile[] unsavedFiles) 
```
Minor style issue - CamelCase method name, space before the (

mhutch replied:
(there are other instances of this, I won't comment them all)

mhutch replied:
And you can ignore the C# naming conventions for the DllImports - a 1:1 mapping from C to C# is fine.

gudaol replied:
Fixing

mhutch commented on CBinding/CBinding.csproj:
```diff
@@ -106,57 +109,75 @@
     </Reference>
-    <Reference Include="System.Composition.Convention">
-      <HintPath>..\packages\Microsoft.Composition.1.0.27\lib\portable-net45+win8+wp8+wpa81\System.Composition.Convention.dll</HintPath>
+    <Reference Include="MonoDevelop.Deployment.Linux">
```
You definitely shouldn't need to reference this addin.

gudaol replied:
its a legacy reference, removing

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,528 @@
+			lock (SyncRoot) {
+				proj.db.Reset (fileName);
+				CXCursor TUcursor = clang.getTranslationUnitCursor (TU);
+				TranslationUnitParser parser = new TranslationUnitParser (proj.db, fileName, cancellationToken, TUcursor);
```
This is an ideal place to use `var`. Have you tried turning source analysis on?

mhutch replied:
Clarification - the type is obvious in cases like `var f = new Foo()`. I wouldn't use var in cases like `Foo f = DoStuff()`.

gudaol replied:
okay, replacing them

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,528 @@
+		/// <param name = "editor">
+		/// A <see cref="CTextEditorExtension"/>: the editor, which edits the document. Contains caret positions, etc.
+		/// </param>
+		public IntPtr codeComplete (
```
It's not obvious what the IntPtr's ownership and thread safety is here. Would be good to address in a comment, at least.

gudaol replied:
added

mhutch commented on CBinding/CLang/CLangManager.cs:
```diff
@@ -0,0 +1,528 @@
+				string name = fileName;
+				CXTranslationUnit TU = TranslationUnits [name];
+				string complete_filename = fileName;
+				uint complete_line = Convert.ToUInt32 (completionContext.TriggerLine);
```
Can't you just cast this?

mhutch replied:
Note - this applies to all the other places, I'm not going to comment them all.

gudaol replied:
replaced with cast

mhutch commented on CBinding/Generated.cs:
```diff
@@ -0,0 +1,2055 @@
+
+    public enum CXErrorCode : uint
+    {
+        @CXError_Success = 0,
```
I know it's not your code but it would be nice to strip the prefix from these enums so it becomes e.g. `CXErrorCode.Success` instead of `CXErrorCode.CXError_Success`. Should be a pretty easy change in the generator. Not a big deal though, could be done later.

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+						unsavedFiles.Add (unsavedFile);
+					}
 				}
+				string operatorPattern = "operator\\s*(\\+|\\-|\\*|\\/|\\%|\\^|\\&|\\||\\~|\\!|\\=|\\<|\\>|\\(\\s*\\)|\\[\\s*\\]|new|delete)";
```
Is this used?

gudaol replied:
yes, to filter operator overloads from code complete results

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+			if (shouldCompleteOn(completionChar)) {
+				CProject project = DocumentContext.Project as CProject;
+				unsavedFiles = new List<CXUnsavedFile> ();
+				foreach (Document openDocument in IdeApp.Workbench.Documents) {
```
You could probably gain some perf by caching the unsaved files until they actually change.

mhutch replied:
For example, you could keep a list of unsaved files on the TextEditorExtension, clear it when the editor loses focus, and update it when the editor gains focus.

This can be done later however, not necessary for the merge.

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_UnaryOperator
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_BinaryOperator
+								|| resultItem.CursorKind == CXCursorKind.CXCursor_CompoundAssignOperator
+								|| System.Text.RegularExpressions.Regex.IsMatch (realstring, operatorPattern)
```
Compiling and retaining the regex instance is better than recreating it every time.

gudaol replied:
replaced

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+							if (clang.getCompletionChunkKind(completionString.Pointer, j) != CXCompletionChunkKind.CXCompletionChunk_TypedText)
+								continue;
+							uint priority = clang.getCompletionPriority (completionString.Pointer);
+							CXString cxstring = clang.getCompletionChunkText(completionString.Pointer, j);
```
This code is very dense and difficult to read, it would be better with some helpers (and empty lines).

For example:

```
IEnumerable<ClangCompletionUnit> GetCompletionUnits (CXCompletionResult result)
{
    var completionString = new CXCompletionString (result.CompletionString);
    uint chunkCount = clang.getNumCompletionChunks (completionString.Pointer);

    for (uint j = 0; j < chunkCount ; j++) {
        if (clang.getCompletionChunkKind (completionString.Pointer, j) != CXCompletionChunkKind.TypedText)
            continue;

        switch (resultItem.CursorKind) {
            case CXCursorKind.Destructor:
            case CXCursorKind.UnaryOperator:
            case CXCursorKind.BinaryOperator:
            case CXCursorKind.CompoundAssignOperator:
                continue;
            case CXCursorKind.ClassDecl:
            case CXCursorKind.StructDecl:
                if (fieldOrMethodMode)
                     continue;
                break;
        }
        string completionText = ConvertString (clang.getCompletionChunkText(completionString.Pointer, j));

        if (operatorRegex.IsMatch (completionText))
            continue;

        uint priority = clang.getCompletionPriority (completionString.Pointer);
        yield return new ClangCompletionUnit (result, completionText , priority);
}

string ConvertString (CXString cxstring)
{
    string s = Marshal.PtrToStringAnsi (clang.getCString (cxstring));
    clang.disposeString (cxstring);
    return s;
}
```

Note that I've also moved operations around for efficiency so that for example we skip marshalling the completion text when we're skipping it due to the cursor type.

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+		/// <param name="completionContext">Completion context.</param>
+		/// <param name="completionChar">Completion char.</param>
+		/// <param name="token">Token.</param>
 		public override Task<ICompletionDataList> HandleCodeCompletionAsync (CodeCompletionContext completionContext, char completionChar, CancellationToken token = default(CancellationToken))
```
It would be good to make this properly async but that can be saved for after the PR is merged.

mhutch commented on CBinding/Gui/CTextEditorExtension.cs:
```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
-					return Task.FromResult ((ICompletionDataList)pair.Value (this, pair.Key, itemName));
+			ICompletionDataList list = new CompletionDataList ();
+			if (shouldCompleteOn(completionChar)) {
+				CProject project = DocumentContext.Project as CProject;
```
Don't use an `as` cast when you expect it to always succeed. `as`returns null if the cast fails, so could cause a `NullReferenceException` when you access the object's member. A 'hard' cast will give an `InvalidCastException` if it fails:

```
var project = (CProject)DocumentContext.Project;
```

gudaol replied:
replaced them

mhutch commented on CBinding/Parser/CDocumentParser.cs:
```diff
@@ -25,58 +25,111 @@
+			foreach (Document openDocument in MonoDevelop.Ide.IdeApp.Workbench.Documents) {
+				if (openDocument.IsDirty) {
+					CXUnsavedFile unsavedFile = new CXUnsavedFile ();
+					unsavedFile.Filename = openDocument.FileName;
```
Object initializer would help readability

mhutch commented on CBinding/Parser/CDocumentParser.cs:
```diff
@@ -25,58 +25,111 @@
+			Project = proj;
+			Manager = proj.cLangManager;
+			unsavedFiles = new List<CXUnsavedFile> ();
+			foreach (Document openDocument in MonoDevelop.Ide.IdeApp.Workbench.Documents) {
```
This needs to run on the UI thread. This should be made clear in the ctor doc comments.

mhutch commented on CBinding/Parser/CDocumentParser.cs:
```diff
@@ -25,58 +25,111 @@
+					unsavedFile.Filename = openDocument.FileName;
+					unsavedFile.Length = openDocument.Editor.Text.Length;
+					unsavedFile.Contents = openDocument.Editor.Text;
+					if (Project.BOMPresentInFile [openDocument.FileName]) {
```
This is the second place I've seen this exact code, and it's not immediately obvious why it's necessary. I would factor it out to a helper method with a descriptive name and comment.

mhutch commented on CBinding/Parser/CDocumentParser.cs:
```diff
@@ -25,58 +25,111 @@
+	/// <summary>
+	/// clang-based document parser helper
+	/// </summary>
+	public class CDocumentParser:  TypeSystemParser
```
Suggest moving this class to its own file.

gudaol replied:
moved

mhutch commented on CBinding/Parser/ClangProjectSymbolDatabase.cs:
```diff
@@ -0,0 +1,442 @@
+					db.Add (file, new ClangFileSymbolDatabase(project, file));
+				}
+				db [file].AddToDatabase (cursor, global);
+			} catch (ArgumentException) {
```
ArgumentException generally implies a bug in code, you shouldn't suppress it.

mhutch commented on CBinding/Parser/DataProvider.cs:
```diff
@@ -27,109 +28,159 @@
+		string completion_string;
+		
+		public CompletionData (CXCompletionResult item, string dataString){
+			if (item.CursorKind == CXCursorKind.CXCursor_ClassDecl) {
```
I would use a switch here.

gudaol replied:
converted

mhutch commented on CBinding/Parser/DataProvider.cs:
```diff
@@ -27,109 +28,159 @@
 
+		public CompletionData (Symbol item)
+		{
+			if (item is Class) {
```
And here.

gudaol replied:
Is switching on Type.ToString() good?

mhutch commented on CBinding/Parser/DataProvider.cs:
```diff
@@ -27,109 +28,159 @@
+		IconId image;
+		string text;
+		string description;
+		string completion_string;
```
Use camelCase for fields.

mhutch commented on CBinding/Parser/DataWrapper.cs:
```diff
@@ -30,56 +30,67 @@
 	sealed class DataWrapper : ParameterHintingData
 	{
-		readonly Function f;
+		OverloadCandidate Overload {
```
This can be a one-liner.

Nice use of C# 6 btw, but since it's private IMO a readonly field would be cleaner than a get-only property.

gudaol replied:
changed

mhutch commented on CBinding/Parser/OverloadCandidate.cs:
```diff
@@ -0,0 +1,28 @@
+			set;
+		}
+
+		public OverloadCandidate(string returns, string name, List<string> parameters){
```
More code styles - { for methods goes on new line, space before (

mhutch commented on CBinding/Parser/Symbol.cs:
```diff
@@ -0,0 +1,204 @@
+	/// </summary>
+	public class Symbol
+	{
+		protected CProject project;
```
Instead of protected backing fields why not convert the properties to use auto properties with a protected set?

gudaol replied:
converted

mhutch commented on CBinding/Parser/Symbol.cs:
```diff
@@ -0,0 +1,204 @@
+					default:
+						return new Symbol (project, fileName, parent, IsGlobal);
+					}
+				} catch (Exception ex){
```
catch-and-ignore is bad. You should at least have a log statement and a comment explaining why.

mhutch commented on CBinding/Parser/Symbol.cs:
```diff
@@ -0,0 +1,204 @@
+			}
+		}
+
+		public string USR {
```
.NET convention is for acronyms/initialisms with 3 letters or more to have the letters following the first to be lowercased. e.g. URLScheme would be UrlScheme in PascalCase and urlScheme in camelCase.

mhutch commented on CBinding/Parser/TranslationUnitParser.cs:
```diff
@@ -0,0 +1,39 @@
+	/// Translation unit parser. Can traverse the translation unit's AST in clang_visitChildren with it's Visit method.
+	/// Builds the symbol database associated with a project.
+	/// </summary>
+	public class TranslationUnitParser{
```
{ for class should be on new line

mhutch commented on CBinding/Parser/TranslationUnitParser.cs:
```diff
@@ -0,0 +1,39 @@
+			if (token.IsCancellationRequested) {
+				return CXChildVisitResult.CXChildVisit_Break;
+			}
+			if (TUCursor.Equals (parent)) {
```
You can simplify this and make it more readable:

```
bool global = TUCursor.Equals (parent);
db.AddToDatabase (file, cursor, global);
```

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -68,15 +67,18 @@ public enum CProjectCommands {
 		
-		[ItemProperty("Target")]
-		CBinding.CompileTarget target = CBinding.CompileTarget.Bin;
+		[ItemProperty ("OutputType")]
```
Doesn't this break existing projects?

gudaol replied:
ugh, i will test it!

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -68,15 +67,18 @@ public enum CProjectCommands {
+		public CLangManager cLangManager;
+		public ClangProjectSymbolDatabase db;
 		
-    	private ProjectPackageCollection packages = new ProjectPackageCollection ();
```
tabs, not spaces

gudaol replied:
sorry these packagecollection and alike are legacy of previous CBinding, will remove them completely when merged into Elsayed's CMakeProject

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -90,118 +92,135 @@ public class CProject : Project, IDeployable
-		}
-		
-		public CProject ()
+		public Dictionary<string, bool> BOMPresentInFile = new Dictionary<string, bool>();
```
Avoid public fields as it breaks encapsulation. A method `public bool IsBomPresentInFile (string filename)` would be better.

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -229,6 +253,10 @@ public void WriteMDPkgPackage (ConfigurationSelector configuration)
 			
 			CProjectConfiguration config = (CProjectConfiguration)GetConfiguration (configuration);
+			while (config == null) {
+				Thread.Sleep (20);
```
This looks like it's hack to fix a race. It should at the very least have a FIXME comment. Fixing the race for real would be much better.

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -437,50 +524,81 @@ protected override SolutionItemConfiguration OnCreateConfiguration (string name,
 		}
-		
+
+		/// <summary>
```
No need to write doc comments on overrides, the docs will be pulled from the base declaration.

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -437,50 +524,81 @@ protected override SolutionItemConfiguration OnCreateConfiguration (string name,
+
 				if (e.ProjectFile.BuildAction == BuildAction.Compile)
-					TagDatabaseManager.Instance.UpdateFileTags (this, e.ProjectFile.Name);
+					ThreadPool.QueueUserWorkItem (o => {
```
This looks really hacky, it's going to have thread safety and race issues. Would be much better to use the type system service's parser queue.

mhutch commented on CBinding/Project/CProjectConfiguration.cs:
```diff
@@ -57,50 +68,83 @@ public enum WarningLevel {
 		[ItemProperty ("Lib", Scope = "*", ValueType = typeof(string))]
-    	private ArrayList libs = new ArrayList ();
-		
+    	ArrayList libs = new ArrayList ();
```
tabs/spaces

mhutch commented on CBinding/Project/CProjectConfiguration.cs:
```diff
@@ -57,50 +68,83 @@ public enum WarningLevel {
 		[ItemProperty ("WarningLevel", DefaultValue=WarningLevel.Normal)]
-		private WarningLevel warning_level = WarningLevel.Normal;
-		
+		WarningLevel warning_level = WarningLevel.Normal;
```
camelCase

mhutch commented on CBinding/Project/CProjectConfiguration.cs:
```diff
@@ -120,15 +168,15 @@ public class CProjectConfiguration : ProjectConfiguration
-				case CompileTarget.Bin:
+				case CompileTarget.Exe:
 					break;
-				case CompileTarget.StaticLibrary:
```
This are backwards. A .NET "module" is more like a C/C++ static library, and a .NET "library" is a shared library.

mhutch commented on CBinding/Project/CProjectConfiguration.cs:
```diff
@@ -57,50 +68,83 @@ public enum WarningLevel {
+		/// Specifies if warnings should be treated as errors or not.
+		/// </summary>
 		[ItemProperty ("WarningsAsErrors", DefaultValue=false)]
-		private bool warnings_as_errors = false;
```
These could probably all be turned into auto properties.

gudaol replied:
sorry, legacy code, changing it

mhutch commented on CBinding/Refactoring/RefactoryCommands.cs:
```diff
@@ -0,0 +1,13 @@
+	/// </summary>
+	public enum RefactoryCommands
+	{
+		GotoDefinition,
```
Should re-use the IDE's command ID, not declare a new one.

gudaol replied:
which one should i use? MonoDevelop.Refactoring.RefactoryCommands does not contain a GoToDefinition

mhutch replied:
Ah, my bad. I forgot C/C++ has a distinction between declaration/definition.

mhutch commented on CBinding/Project/CProject.cs:
```diff
@@ -68,15 +67,18 @@ public enum CProjectCommands {
+		[ItemProperty ("OutputType")]
+		CompileTarget target = CompileTarget.Exe;
+
+		public CLangManager cLangManager;
```
Avoid public fields, especially non-readonly.

mhutch commented on CBinding/Parser/ClangProjectSymbolDatabase.cs:
```diff
@@ -0,0 +1,442 @@
+	/// <summary>
+	/// Symbol database belonging to a project
+	/// </summary>
+	public class ClangProjectSymbolDatabase {
```
Longer term this needs to be serializable so it doesn't have to be rebuilt every time. It could probably be built using the clang_index\* functions.

Commit:
PR requested changes
Cleaned up code style (camelCase, PascalCase, var, simplified if-s, switches, Convert -> cast, compiled regex, initializers, etc.)
Resolved data race in CProject.OnFileAddedToProject
Added helper methods where requested/needed
Removed prefix from enums in generated code
Most of fields -> auto-properties
Removed Deployment references

Commit:
Removed unneeded wait (data race fixed in previous
... commit)

Commit:
Fixed references
